docs/design/persistent-storage.md

```diff
@@ -0,0 +1,222 @@
+
+
+
+#### Pod to specify storage requirements
+
+```
+
+	Pod.Spec.Volumes.Volume 
+			Name
+			VolumeSource -- to be bound with available storage from the pool
+			PersistenceRequirements
+				Mode				ReadWriteOnce | ReadOnlyMany
+				RetentionPolicy 	StoreRetentionPolicy (RetainOnDelete|RecycleOnDelete)				Resources			ResourcesList (size, iops, throughput)
+				Selector			Labels to identify the PersistentStore backing implementation
+
+```
+
+`PersistenceRequirements` are the capabilities and performance characteristics a pod author can request and rely on for the entire lifecycle of the volume.  Requests for storage that do not match available storage will go unfulfilled.
+
+
+#### Storage discovery 
+
+An API and CLI is needed to view available and in-use stores by namespace.
+
+**Use Case**:  Susie needs to crunch some data and serve it afterwards.  A single pod to create data is sufficient, but she estimates needing 10 web front ends to serve the data afterwards.  For this use case, the data does not change once created.  Susie first looks to see what is available in the cluster.
+
+```
+$ cluster/kubectl.sh get storage
+
+NAME                LABELS              MODES               RESOURCES
+_pv_vol_01          name=my-data        RWO,ROX             size=5gb,iops=30
+_pv_vol_02                              RWO,ROX             size=5gb,iops=30
+_pv_vol_03          name=other-data     RWO,ROX             size=25gb,iops=30
+
+```
+
+#### Request persistent storage for a pod
+
+Susie is happy to see storage that can be mounted many times (ReadOnlyMany).  That satisfies her load balancer requirement.  Susie creates an pod image with her number crunching code and requests persistence for her volume.
+
+Susie's YAML looks something like this for the pod that creates her data:
+
+
+```yaml
+
+id: creator-of-data
+kind: Pod
+namespace: ns
+apiVersion: v1beta1
+desiredState:
+  manifest:
+    version: v1beta1
+    id: creator-of-data
+    containers:
+      - name: numbercruncher
+        image: ns/makesomdata:v1
+        command:
+          - /run.sh
+        volumeMounts:
+          - name: data
+            mountPath: /some_data
+    volumes:
+      - name: data
+        persistenceRequirements:
+        	availableModes: readwriteonce | readonlymany
+        	currentMode: readwriteonce
+        	retentionPolicy: retainondelete
+        	resources:
+        		- size: 5
+        	selector:
+        		- name: my-persistent-data
+labels:
+  name: creator-of-data
+
+```
+
+#### Behind the curtain
+
+`PersistentStoreBinder` is watching for new pods with `PersistenceRequirements`.  The binder compares the request to all available stores and binds it to the closest match.
+
+To bind:
+
+* Set PersistentStore.Status.PodRef = pod that originally requested persistence
+* Set PersistentStore.Labels = PersistenceRequirements.Selector
```

Comment:
So you're trying to cut out one user step, but you still have a notion of a claim object, you've just overlaid it on the offer object.  What's awkward to me is that you try to use the same specification for "allocate a PS for me" and "re-attach a PS for me" - but those are very different things!

Getting an empty volume when you expected to get something you had previously prepared seems like a Bad Thing.

What if we considered something like:

* Adam Admin creates a PersistentStore object of type PD, named data-depot-1 in the root of the API (no namespace).

* Beula creates a Pod in namespace nifty-network, with a volume of type "NewPersistent", with fields like claimName=nifty-data, claimLabels, and requirements.

* A PersistentStoreClaim object is created in the Beula's nifty-network namespace, named nifty-data, with labels as specced before.

* The PSBinder finds a PSController that owns a PS that could satisfy the requirements and binds that PS to the PSClaim

* The pod is scheduled and runs to completion.

* The PSClaim still exists.

* Beula creates another Pod in namespace nifty-network, with a volume of type "Persistent", with fields like claimName=nifty-data, claimLabels, and requirements.

* The PSClaim object is resolved and the pod is scheduled.

* When the pod is deleted, the PSClaim still exists.

* Beula gets a nag email from Adam Admin saying that she has a Claim on a PS and if she is done with it, could she please release it.  She does.


I'm not sure the implicit claim creation is worth the effort - if it is a whole distinct step, is that so bad?  This is all about pet storage right?  Pets have names and indentities and owners.